### PR TITLE
Replace jQuery disabling Image URL generating fields with Stimulus `w-rules`/`RulesController`

### PR DESCRIPTION
**Pull request by lb-**
- Initial progress on #11045 (part 2 - replacing `image-url-generator`)
- Adds a new `w-rules` / `RulesController` to support conditional enabled targets based on the data of a form.
- This is a smaller scoped version of the sibling PR #11117 - which focuses on only the `enable` and does not include the ability to conditionally show elements.
- **Update** Originally this was named `w-cond` & `CondController` but after some [discussion](https://github.com/wagtail/wagtail/pull/11202#discussion_r1385664326) we felt that `rules` makes more sense and meets the 'concise name' goal also. The previous branch is up for reference though https://github.com/lb-/wagtail/tree/feature/11045-part-2-replace-image-url-generator-cond-controller-backup

## Details

The current jQuery logic contains a few sets of behaviour, while this does not replace the whole jQuery file it does get us a step closer to removing this and replacing this with more modern patterns.

Once merged, we may want to revisit the idea of building up the Image filter string on the client-side and move that to the server. But we cannot do that without solving the required client-side disabling of fields, hence this PR.

This paves the way for a Wagtail `FieldPanel` approach for dynamic behaviour client-side in the future. See #2172

## Screenshots

Screenshot 2023-11-09 at 6 08 49 pm

Screenshot 2023-11-09 at 6 08 08 pm

## Checks

-   [X] Do the tests still pass?
-   [X] Does the code comply with the style guide?
-   [X] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [X] For front-end changes: Did you test on all of Wagtail’s supported environments?
-   [X] For new features: Has the documentation been updated accordingly?

## Testing

* To access this page, edit an existing image and click the button under the image preview 'URL Generator'.
* The drop-down 'Filter' should conditionally disable/enable the other three fields based on the filter type.

## Reviewing

I have intentionally split out the commits into small reviewable chunks.

1. Adds a JavaScript util `castArray` - very similar to Lodash - https://lodash.com/docs/4.17.15#castArray
2. Adds the initial `RulesController` implementation, supporting `enable` only, with unit tests and a story
3. Finally, updates the HTML to use the new data attribute on the form container, passing the filter data down via attributes to each field and then removal of the jQuery logic that does the same.

### CODE REVIEW

File: client/src/controllers/index.ts
Comment (zerolab) [inline client/src/controllers/index.ts:8]: question: looking at the other controllers, they all seem to be fully named. Is there a reason this is not `ConditionController` ?

Comment (lb-) [reply client/src/controllers/index.ts:8]: I could not think of a better name that was also short.

One of the downsides of a longer name is that this name gets repeated for every single attribute.

For example, comparing the two, we get a much longer value in the DOM. This impacts DOM size and also readability.

```
<input data-w-cond-target="enable" data-action="change->w-cond#something keyup->w-cond#other" />
<input data-w-condition-target="enable" data-action="change->w-condition#something keyup->w-condition#something" />
```

We do say that the controller name should be concise here https://docs.wagtail.org/en/stable/contributing/ui_guidelines.html#how-to-build-a-stimulus-controller


However, I am more than happy to change if this 'make it short' hurts the readability of this.

Comment (lb-) [reply client/src/controllers/index.ts:8]: `w-relate` (related elements)
`w-deps` (dependant elements)
`w-match` (matched elements)
`w-rules` (rules determine the behaviour of fields) - maybe this could be a better option?


Another idea, we could re-label the match attribute to be data-conditions maybe. But I like to separate the controller targets from the way we set the 'rules'.

Comment (zerolab) [reply client/src/controllers/index.ts:8]: No, it makes sense especially in the DOM. Could the controller be more verbosed, but the DOM trigger short? Would `w-rule` work? if not `CondController`/`w-cond` will work (sorry if I am nitpicking)

Comment (zerolab) [reply client/src/controllers/index.ts:8]: Oh, cross-post! `w-rules` sounds like a winner to me at the moment. Before we commit - would we ever get to a position where we'd have a controller for complex rules?

I wonder what other think

Comment (lb-) [reply client/src/controllers/index.ts:8]: I envision we would like to solve this issue https://github.com/wagtail/wagtail/issues/2172 and `RulesPanel` seems a suitable name there. Nothing's stopping us from just making this behaviour a kwarg though. 

`FieldPanel(..., rules=[...]` (or something). 

If we do need to build a new controller for something else where rules is useful as a name for this thing we can just use a different name for that new thing. I'm not aware of anything on our existing Stimulus roadmap that would fit this name though.

The only thing close would be a master config like controller, but it's unlikely we will head in that direction anyway.

Comment (lb-) [reply client/src/controllers/index.ts:8]: Here's what the HTML would look like. In this example we are setting the `beta` field to be enabled only when you enter `'go'` into the `alpha` field. Contrived but should give a sense of things.

```
<fieldset data-controller="w-rules" data-action="change->w-rules#resolve">
  <input type="text" name="alpha" />
  <input type="text" name="beta"  data-w-rules-target="enable" data-match='{"alpha": "go"}' />
</fieldset>
```

An alternative approach where the `data-match` is instead `data-rule` (singular).

```
<fieldset data-controller="w-rules" data-action="change->w-rules#resolve">
  <input type="text" name="alpha" />
  <input type="text" name="beta"  data-w-rules-target="enable" data-rule='{"alpha": "go"}' />
</fieldset>
```

A future enhancement may allow for per-target type rules, so it would be good to consider this in the context of changing this attribute. Both variants listed below.

```
<fieldset data-controller="w-rules" data-action="change->w-rules#resolve">
  <input type="text" name="alpha" />
  <input type="text" name="beta"  data-w-rules-target="enable show" data-match-enable='{"alpha": "go"}' data-match-show='{"alpha": "blah"}'  />
  <input type="text" name="beta"  data-w-rules-target="enable show" data-rule-enable='{"alpha": "go"}' data-rule-show='{"alpha": "go"}' />
</fieldset>
```

Comment (lb-) [reply client/src/controllers/index.ts:8]: No major opinions from the core team. Aside from the eventual Python API that uses these attributes may end up having a more descriptive name. 

I'm still leaning towards rules but will wait for @zerolab or anyone else to confirm.

Comment (zerolab) [reply client/src/controllers/index.ts:8]: I like rules, fwiw

**Comment by lb-** [comment]
@zerolab @thibaudcolas (or others) this PR has been updated as per the discussion to use the `w-rules` naming. I have tested these changes and they work reliably on the image URL generator page. I have updated tests/storybook accordingly also.

**Comment by zerolab** [comment]
This LGTM 🙌🏼, but would be good to get a FE person for a full review

**Comment by lb-** [comment]
@thibaudcolas - please can you keep this on your radar for reviews.

This code has been validated by Dan but needs your or maybe @laymonage to check you are happy with wording/naming.

Once merged in, this creates a great path forward for us to remove a bunch of other ad-hoc jQuery & script tags and will be a key part of the Stimulus migration. Also, this could help us towards a longstanding request for dynamic fields in Wagtail. See https://github.com/wagtail/wagtail/issues/11045 & the description.

I've done my best to keep the scope of this to an absolute minimum but have other functional code ready to go in another PR after this for the dynamic filter/selects.

File: client/src/controllers/RulesController.ts
Comment (laymonage) [inline client/src/controllers/RulesController.ts:129]: I don't see any tests nor usage of this with the `hidden` attr, and the only call to this method (in `resolve()`) has the parameter hardcoded to `disabled`. I know you likely meant for this to be reusable with different attributes to be toggled, e.g. `hidden` here to hide the field instead of disabling it. However, I'm not sure this is even possible with the current implementation, and it ends up somewhat muddying the code for what we currently use it for.

For example, we end up adding special cases with "half-measures" here:

I don't understand why we're special-casing `'disabled'` to use `target.disabled = true` instead of `setAttribute("disabled", "")`, yet we use `removeAttribute()` instead of `target.disabled = false`. I'd rather we stick to one approach for both disabling and enabling (either use `setAttribute(attr, "")` and `removeAttribute(attr)`, or `target.disabled = true` and `target.disabled = false`).

Additionally, AFAIK the `hidden` and `disabled` properties behave similarly, as in:

- setting `.disabled = true`/`.hidden = true` will add a `"disabled"`/`"hidden"` attribute with an empty string `""` as the value
  - this is the same as `setAttribute("disabled", "")`/`setAttribute("hidden", "")`, which would also make the respective property return `true`
- setting `.disabled = false`/`.hidden = false` will remove the `"disabled"`/`"hidden"` attribute (if exists)
  - this is the same as `removeAttribute("disabled")`/`removeAttribute("hidden")`, which would also make the respective property return `false`

So, `disabled` doesn't need special casing any more than `hidden` does.

Also, I'm not sure when the `target.setAttribute(attr, attr);` case would be useful (there is no test case that demonstrates this).

Perhaps these are leftovers from the `w-cond` implementation?

Sorry for asking to trim down a controller's implementation yet again, I know we've had similar discussions in the past. However, without a usage in core or a test to verify this works as expected, it's hard to justify the additional complexity here 🙏

**Review by laymonage — COMMENTED** [review]
Hey LB, thanks for working on this! While it does work for the purpose of the feature, I think we could either simplify the implementation for that purpose only, or ensure the additional features actually work by adding more test cases.

I don't have any strong opinions on the naming, so `RulesController` works for me.

**Comment by lb-** [comment]
Thank you @laymonage - I'll get onto this when I can. Great feedback, I'll look into the superset thing and check if there's another approach or a polyfill. 

Good call about the unused functionality, it's probably a left over from my 'full featured' implementation that I cut down for the PR.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:89]: We probably need to agree on our code styleguide whether we want to use `dataset`. Originally I was using this a lot (when starting with Stimulus), but within the Wagtail codebase I found it's much easier to search through the code for the full attribute name (e.g. for debugging or code spelunking).

I think here, it's pretty straight forward and using `dataset` makes sense but just explaining why I opt for the attribute `getApproach` in most cases.

Comment (laymonage) [reply client/src/controllers/RulesController.ts:89]: @lb- You’re right, I’ve stumbled upon the same issue a few times and have always had to search for the dataset equivalents as well. With that in mind, I think now I’m in favour of using getAttribute.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:109]: My goal here is to allow for `data-rule='{"fav-drink": false}'` to trigger when the field with name `fav-drink` is has a blank (empty string) value. This was me thinking about how it would be intuitive to write these rules.

However, I see your point. We could assume that anyone writing these rules would know that form values can only be strings, but then using `false`, `null`, or `undefined` would basically never work at all.

Unless we do coerce matching, but I feel this makes the complexity in the matcher too much.

For now, I will just remove this smartness, we will trust devs to know that an empty string is the only way to check for 'falsey' values.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:104]: Hmm, actually with the removal of the falsey string generation, I might go ahead with this suggestion. Just let the controller fail if it cannot parse the JSON.

**Comment by lb-** [comment]
@laymonage @thibaudcolas 

I have now done a large refactor of this PR with all the feedback in mind.

1. Remove any logic or structure that considers future enhancements of this controller, we can refactor again later if that's needed. Thanks Sage for the push here, it was complex to understand the logic as it was present in the PR, I think overall this is now in a simpler to reason about state.
2. Remove special handling for select fields (when disabling/enabling), this can be added later when we do future enhancements to this controller.
3. Ensure we handle cases where the same field name can result in multiple values, yes this is for the future but I agree this is a fundamental item to consider. I have added unit tests for this also.
4. Avoid the usage of `data-rule` explicitly, and instead rely on the identifier to construct a rule attribute e.g. (`data-w-rules`). This way we keep in line with the Stimulus practice of not hard-coding attribute values if we can avoid it, allowing this same controller to be used with different identifier registrations.
5. Simplify the string conversion for rule check values, allow `false` to become `'false'` and `null` to become `'null'` etc, so that we have a consistent way to understand the rules by assessing the HTML.
6. Add better handling of the JSON parsing, using Stimulus built in capability to throw errors nicely (see screenshot, the awesome thing is that even if one field's rules are broken, the other fields still work and a helpful error is thrown in the console with the issue).
7. Instead of storing the rules as an object, store as an array, this means less double handling of the data structure and a more consistent TypeScript usage in the code.
8. Completed new manual testing on Safari, Chrome, Firefox, reviewed Storybook works still.

## Updated screenshots

Screenshot 2024-10-09 at 7 27 29 am
Screenshot 2024-10-09 at 7 27 25 am

**Intentional error added to rules to validate it throws correctly**

Screenshot 2024-10-09 at 7 41 34 am

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts]: I've cleaned up this approach now, it's much simpler.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts]: Ok. Finally landed on an approach that's testable and throws an error using the built in Stimulus error handling. 

https://stimulus.hotwired.dev/handbook/installing#error-handling

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts]: This has been updated.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts]: Agreed. I've removed all of this now.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts]: This is helpful feedback, thankyou. 

I've simplified this whole approach, we can review in the future if we need to abstract out to separate methods.

**Comment by thibaudcolas** [comment]
Assigning to @laymonage specifically for review, and let me know if I should as well.

File: client/src/controllers/RulesController.test.js
Comment (laymonage) [inline client/src/controllers/RulesController.test.js:146]: Would `type="checkbox"` be more appropriate for this test? You can't multi-select radio buttons, can you?

File: client/src/controllers/RulesController.ts
Comment (laymonage) [inline client/src/controllers/RulesController.ts:68]: Worth thinking about: this currently matches "any" valid value in the rules. I can see this controller being useful for enforcing "all" the valid values, e.g. requiring a set of fields to be selected in a select multiple widget. There's no such need for it in the codebase currently, but I assume it would be as simple as changing the logic from `.some()` to `.every()` here.

If that's the case, that makes me wonder: should we maybe rename the `data-w-rules` attribute to something like `data-w-rules-any` to match any value, with the possibility of adding `data-w-rules-all` support later? (Naming TBD, could also use `data-w-rules-some` and `data-w-rules-every` if that's preferred. I think `any` and `all` ala Python is more readable though)

I think even if we don't add support for "all" in this PR, `data-w-rules-any` makes it clear that you only expect at an "any" match with the values in the rules, whereas with `data-w-rules` it's not very obvious. The docstring of the controller at the moment also does not make it clear that it's an "any" matching.

**Review by laymonage — COMMENTED** [review]
Thank you, this seems to work well and the code is now so much easier to get through!

I think this is pretty close, but there some suggestions to think about. Thanks again for working on this, and sorry I only got around to reviewing this again now!

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:68]: Yeah good thinking, I'd started down that track (something similar, I'll have to find my older branches) but felt it was too much 'future use'. I think maybe we'll start with a sensible default behaviour and see what use cases come up.

I also plan to enhance to support enable and disable options and was going to use the data attribute to support this. Such as `data-w-rules-enable`, `data-w-rules-disable`, `data-w-rules-show` and `data-w-rules-hide`.

I'll have a think about this and see what could work, I like the idea of the default being that all must match. It feels more intuitive.

Just brainstorming, but we could pass in a special name to the rule to support this. For example, `noindex` or an empty name are not valid input name saccording to the spec. https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#naming-form-controls:-the-name-attribute

We could use this to support something like `data-w-rules='{"":"any",...}'`, `data-w-rules='{"noindex":"any",...}'`.

I was thinking of these directives; `all`, `any`, `non`, `one`. This would probably be abstracted away at the Django widget level with specific kwargs that make sense in python. I've used these names in a different project, three characters and alpha sorting gives a logical default of 'all'.

File: client/src/controllers/RulesController.test.js
Comment (lb-) [reply client/src/controllers/RulesController.test.js:146]: Good call. I'll fix this up.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:49]: Hmm. Good point, I'll take another look at this.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:80]: Yeah, I'll go this direction, I'll add a lint ignore rule for this file to allow mutations on this attribute.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:30]:
```suggestion
  /** True if there is at least one enable target, used to ensure rules do not run if not needed. */
```

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:67]: Nice. I'll give this a go and update unit tests accordingly.

File: client/src/controllers/RulesController.ts
Comment (lb-) [inline client/src/controllers/RulesController.ts:83]: I might also pass something useful about what actually resolved and what target was changed.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:68]: OK, this suggestion is one level lower than my reply is about. As it's not the ruleset that this line relates to but the value matching for a field.

In which case, I actually think we should leave it as is for now. This change is a bit painful to make and in my mind makes it harder for the case where you want to support more flexible matching.

I feel that this PR as is has a solid enough starting point for us to build on, we may want to consider how the value matching works later. We could support inner objects within the value matching, not just an array but maybe an object that allows more complex things.

Thanks @laymonage  though, this is good input, I will do my best to sum this up into a new issue for future enhancements both at the ruleset (per element) level and at the field value matching. I also think we will need to consider complex cases such as streamfield eventually.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts:67]: OK, I went a slightly different way here, but have added a test and refined the 'default' value (now an empty array) to support this.

**Comment by lb-** [comment]
@laymonage this is now ready for another review.

File: wagtail/images/forms.py
Comment (laymonage) [inline wagtail/images/forms.py:217]: I guess this is too late, but now I do wonder whether it makes more sense to put the controller on these elements instead of the form (and get the form values via [`element.form`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/form)).

The problem with hooking them as targets is that in order to use these features, you have to rely on the `<form>` having `data-controller="w-rules"` and its `data-action` predefined. In some cases it may not be easy to override the `<form>` element (e.g. in a generic template).

But I guess the difficulty in that approach is that it's harder to listen to the `change`/`keyup` events unless you attach event listeners to the form element inside the controller code (instead of using `data-action`). Also might not be a good idea to have one event listener for each element with rules...

Anyway, I think I'm overthinking this – there's only one such use case within Wagtail so far anyway. The other potential reuse is the `ImageBlock` for handling the "is decorative" checkbox, but I guess if we put the controller on the `<form>` in the page and generic templates, we should be OK...

**Review by laymonage — APPROVED** [review]
The code is now very easy to read, thank you!

I have one thought but I think it's too late to reconsider now, and I haven't fully explored whether that would be a good idea anyway. I'll leave it for you to merge in case my comment changes anything, but otherwise this looks good to me!

File: wagtail/images/forms.py
Comment (lb-) [reply wagtail/images/forms.py:217]: Thanks @laymonage - it's funny you say that, I actually did something similar in my earliest POCs.

I think for now we can leave as is, but I agree that it may be better to do go down a different track in the future. At the time though I assumed we would not want to  add global listeners and was hoping to avoid controllers adding/removing their own listeners if needed.

In saying that, it would only be a small change to support something like this, I will make this small adjustment when merging. Essentially allowing the `this.element` to be any input-like element and I will find the nearest `form`.

```html
<input type="text" name="foo" data-controller="w-rules" data-action="change@document->w-rules#resolve keyup@document->w-rules#resolve" data-w-rules-target="enable" data-w-rules='{"other": ["a", "b", "c"]}' />
```

After this, the next step would be to probably add an action filter to allow us to filter events via the current element's form. But I will not do that on this PR.


```html
<input type="text" name="foo" data-controller="w-rules" data-action="change@document->w-rules#resolve:form keyup@document->w-rules#resolve:form" data-w-rules-target="enable" data-w-rules='{"other": ["a", "b", "c"]}' />
```

Alternatively, we could just add a value to allow the listeners to be hooked up by the controller.

File: client/src/controllers/RulesController.ts
Comment (lb-) [reply client/src/controllers/RulesController.ts]: I have added a reminder about this on the Stimulus migration project - will assess when we have more momentum on known use cases first.

https://github.com/orgs/wagtail/projects/14/views/1?pane=issue&itemId=89429564

Comment (lb-) [reply client/src/controllers/RulesController.ts]: PR up that includes some of these improvements, without going to deep into 'future' use cases though.
https://github.com/wagtail/wagtail/pull/12667

**Pull request merged**
